sys/pool_alloc.c

```diff
@@ -0,0 +1,235 @@
+  bitstr_t *bitmap = slab->ph_bitmap;
+  int i = 0;
+  bit_ffc(bitmap, slab->ph_ntotal, &i);
+  pool_item_t *found_pi = GET_PI_AT_IDX(slab, i, size);
+  if (found_pi->pi_guard_number != PI_MAGIC_WORD) {
+    panic("memory corruption at item 0x%lx\n", (unsigned long)found_pi);
+  }
+  slab->ph_nused++;
+  slab->ph_nfree--;
+  bit_set(bitmap, i);
+  PALLOC_DEBUG_PRINT("Allocated an item (0x%lx) at slab 0x%lx, index %d\n",
+                     (unsigned long)found_pi->pi_data,
+                     (unsigned long)found_pi->pi_slab, i);
+  return found_pi->pi_data;
+}
+
+void pool_init(pool_t *pool, size_t size, void (*constructor)(void *, size_t),
+               void (*destructor)(void *, size_t)) {
+  /*
+  init object and give it 1-page cache;
+  */
+  PALLOC_DEBUG_PRINT("Entering pool_init\n");
+  size = ALIGN_TO_BINARY_POW(size, 2); /* Align to 32-bit (4 byte) word */
+  LIST_INIT(&pool->pp_empty_slabs);
+  LIST_INIT(&pool->pp_full_slabs);
+  LIST_INIT(&pool->pp_part_slabs);
+  pool_slab_t *first_slab = create_slab(size, pool, constructor);
+  LIST_INSERT_HEAD(&pool->pp_empty_slabs, first_slab, ph_slablist);
+  pool->pp_constructor = constructor;
+  pool->pp_destructor = destructor;
+  pool->pp_itemsize = size;
+  pool->pp_nslabs = 1;
+  pool->pp_nitems = first_slab->ph_ntotal;
+  PALLOC_DEBUG_PRINT("Initialized new pool at 0x%lx\n", (unsigned long)pool);
+}
+
+void pool_destroy(pool_t *pool) {
+  /*
+  destroy the objects;
+  free the underlying memory;
+  */
+
+  PALLOC_DEBUG_PRINT("Entering pool_destroy\n");
+
+  unsigned long *tmp = (unsigned long *)pool;
+
+  for (size_t i = 0; i < sizeof(pool_t) / sizeof(unsigned long); i++) {
+    if (tmp[i] == PP_FREED_WORD) {
+      panic("double free at pool 0x%lx\n", (unsigned long)pool);
+    }
+  }
+
+  PALLOC_DEBUG_PRINT("Destroying empty slabs\n");
+  pool_slab_t *it = LIST_FIRST(&pool->pp_empty_slabs);
+  while (it) {
+    pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+    LIST_REMOVE(it, ph_slablist);
+    destroy_slab(it, pool);
+    it = next;
+  }
+
+  PALLOC_DEBUG_PRINT("Destroying partially filled slabs\n");
+  it = LIST_FIRST(&pool->pp_part_slabs);
+  while (it) {
+    pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+    LIST_REMOVE(it, ph_slablist);
+    destroy_slab(it, pool);
+    it = next;
+  }
+
+  PALLOC_DEBUG_PRINT("Destroying full slabs\n");
+  it = LIST_FIRST(&pool->pp_full_slabs);
+  while (it) {
+    pool_slab_t *next = LIST_NEXT(it, ph_slablist);
+    LIST_REMOVE(it, ph_slablist);
+    destroy_slab(it, pool);
+    it = next;
+  }
+
+  for (size_t i = 0; i < sizeof(pool_t) / sizeof(unsigned long); i++) {
+    tmp[i] = PP_FREED_WORD;
+  }
+  PALLOC_DEBUG_PRINT("Destroyed pool at 0x%lx\n", (unsigned long)pool);
+}
+
+void *
+pool_alloc(pool_t *pool,
+           __unused uint32_t flags) { /* TODO: implement different flags */
+  /*
+  if (there’s an object in the cache)
+      take it (no construction required);
+  else {
+      allocate memory;
+      construct the object;
```

Comment:
Is this comment accurate? No object construction happens in this function. Maybe you meant slab creation? Anyway, *interleaving* the code with comments makes it simpler to read than providing the outline at the beginning of some functions. This way, when one reads thorough the code, the comments explain which part of the code is responsible for which part of your logic. A general outline before the actual code is only helpful if the code is very complex and the outline significantly simplifies the code.